### PR TITLE
chore: try to reintroduce Python 3.12, now that NumPy has released a beta.

### PR DESCRIPTION
**Pull request by jpivarski**
Here's the first test (on this branch, not deploying to PyPI):

https://github.com/scikit-hep/awkward/actions/runs/5857811221

### CODE REVIEW

**Comment by jpivarski** [comment]
Currently, the error is due to the `cmake` package on PyPI. The [Linux Python 3.12 test](https://github.com/scikit-hep/awkward/actions/runs/5900302218/job/16004317102?pr=2644) fails because

```
... (lots of output, ending with) ...
  pip._vendor.pyproject_hooks._impl.BackendUnavailable: Traceback (most recent call last):
    File "/opt/hostedtoolcache/Python/3.12.0-rc.1/x64/lib/python3.12/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 77, in _build_backend
      obj = import_module(mod_path)
            ^^^^^^^^^^^^^^^^^^^^^^^
    File "/opt/hostedtoolcache/Python/3.12.0-rc.1/x64/lib/python3.12/importlib/__init__.py", line 90, in import_module
      return _bootstrap._gcd_import(name[level:], package, level)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "<frozen importlib._bootstrap>", line 1293, in _gcd_import
    File "<frozen importlib._bootstrap>", line 1266, in _find_and_load
    File "<frozen importlib._bootstrap>", line 1216, in _find_and_load_unlocked
    File "<frozen importlib._bootstrap>", line 400, in _call_with_frames_removed
    File "<frozen importlib._bootstrap>", line 1293, in _gcd_import
    File "<frozen importlib._bootstrap>", line 1266, in _find_and_load
    File "<frozen importlib._bootstrap>", line 1237, in _find_and_load_unlocked
    File "<frozen importlib._bootstrap>", line 841, in _load_unlocked
    File "<frozen importlib._bootstrap_external>", line 994, in exec_module
    File "<frozen importlib._bootstrap>", line 400, in _call_with_frames_removed
    File "/tmp/pip-build-env-1e13_weg/overlay/lib/python3.12/site-packages/setuptools/__init__.py", line 10, in <module>
      import distutils.core
  ModuleNotFoundError: No module named 'distutils'

  error: subprocess-exited-with-error
  
  × pip subprocess to install build dependencies did not run successfully.
  │ exit code: 2
  ╰─> See above for output.
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  full command: /opt/hostedtoolcache/Python/3.12.0-rc.1/x64/bin/python /opt/hostedtoolcache/Python/3.12.0-rc.1/x64/lib/python3.12/site-packages/pip/__pip-runner__.py install --ignore-installed --no-user --prefix /tmp/pip-build-env-opv7ul0k/overlay --no-warn-script-location --no-binary :none: --only-binary cmake -i https://pypi.org/simple -- 'cython >= 0.29.31' 'oldest-supported-numpy>=0.14' setuptools_scm 'setuptools >= 40.1.0' wheel
  cwd: [inherit]
  Installing build dependencies: finished with status 'error'
error: subprocess-exited-with-error

× pip subprocess to install build dependencies did not run successfully.
│ exit code: 2
```

That's a long pip-install command, but I see `cmake` as the only package mentioned among all of the installation flags.

In the `cmake` pip package's setup.py file, I see

https://github.com/scikit-build/cmake-python-distributions/blob/0acab486a2d7b4ae06b9c39a5d178402ee5b0afe/setup.py#L6

trying to import `distutils`. I also looked at the source package for the [most recent release of the `cmake` package on PyPI](https://pypi.org/project/cmake/3.27.2/#files) and the

```
from distutils.text_file import TextFile
```

line is in that setup.py, too.

I don't see any [open issues on the `cmake` repository](https://github.com/scikit-build/cmake-python-distributions/issues) about "3.12" or "distutils", so I'll report this there unless @henryiii thinks this couldn't be related for some reason. (We are, after all, trying to get a binary-only `cmake` package, and the setup.py file is not in the binary wheels.)

**Comment by agoose77** [comment]
I don't *think* this is CMake. When I was digging, it looked like maybe `oldest-supported-numpy` was pulling in a NumPy sdist. I was unsure as to why this happens, as we also pull in the pre-release. I wondered if the sdist needed to be build in order for pip to find the metadata, but that seems unlikely.

**Comment by jpivarski** [comment]
Awkward 2.4.0 will not support Python 3.12, but the next one probably will.

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -192,12 +201,12 @@ jobs:
         include:
           - python-version: '3.8'
             numpy-package: "numpy==1.18.0"
-            pyarrow-package: "pyarrow"
-          - python-version: '3.8'
-            numpy-package: "numpy"
             pyarrow-package: "pyarrow==7.0.0"
+          - python-version: '3.12'
+            numpy-package: "numpy>=1.26.0b1"
+            pyarrow-package: "pyarrow;python_version<'3.12'"
```
Comment (agoose77) [inline .github/workflows/test.yml:210]: Disable pyarrow for 3.12 without breaking shell

**Comment by agoose77** [comment]
@jpivarski can you give this a once-over? I'm happy to merge now.

**Review by jpivarski — COMMENTED** [review]
I think it looks good and can go ahead before we hear back from @henryiii on Slack about NumPy's thin wheels.

I apparently can't press the "approve" button on my phone, but I approve this.

The two tests that haven't run are the ones with minimal NumPy and minimal pyarrow. The fact that this was merged with main and they're not running could be a sign that they were accidentally dropped in a merger (since you're touching the same files). Could you check on that, to make sure they don't get lost when you merge this into main?

**Comment by henryiii** [comment]
You can approve from a phone as long as you are it the code view part (on web or GitHub's App).

**Comment by agoose77** [comment]
@jpivarski this PR fuses the two separate NumPy and pyarrow lower-bound tests to run them at the same time. I don't think this is essential, so I'll revert it such that we don't need to change our required tests.

**Pull request merged**